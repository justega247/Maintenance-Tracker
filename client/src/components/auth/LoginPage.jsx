import React, { Component } from 'react';
import { Link, Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';
import Button from '../common/Button';
import ValidateUser from '../../utils/authValidation';
import { frontendRoutes } from '../../constants/routes';
import { startUserSignIn } from '../../actions/auth';

export class LoginPage extends Component {
  state = {
    username: '',
    password: '',
    errors: {},
  };

  onChange = (event) => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  };


  onSubmit = (event) => {
    event.preventDefault();

    const { errors, hasErrors } = ValidateUser.signInDataValidation(this.state);
    this.setState(() => ({
      errors,
    }));

    if (!hasErrors) {
      const { username, password } = this.state;
      const { startUserLogin } = this.props;
      startUserLogin({ username, password });
    }
  };


  render() {
    const { errors } = this.state;
    const { auth, userId } = this.props;
    const RedirectionRoute = userId === 1 ? `${frontendRoutes.ADMIN_DASHBOARD}` : `${frontendRoutes.USER_DASHBOARD}`;
    if (auth) {
      return <Redirect to={RedirectionRoute} />;
    }

    return (
      <div>
        <div>
          <form onSubmit={this.onSubmit} className="auth__form">
            <div className="auth__body">
              <div>
                <label htmlFor="username" className="label">
                  Username:
                  <div>
                    <input
                      type="text"
                      name="username"
                      placeholder="enter your username"
                      className="form__input"
                      value={this.state.username}
                      onChange={this.onChange}
                    />
                  </div>
                </label>
                {errors.username &&
                  errors.username.map(error => (
                    <div className="form__error" key={error}>
                      {error}
                    </div>
                  ))}
              </div>
              <div>
                <label htmlFor="password" className="label">
                  Password:
                  <div>
                    <input
                      type="password"
                      name="password"
                      placeholder="enter your password"
                      className="form__input"
                      value={this.state.password}
                      onChange={this.onChange}
                    />
                  </div>
                </label>
                {errors.password &&
                  errors.password.map(error => (
                    <div className="form__error" key={error}>
                      {error}
                    </div>
                  ))}
              </div>
              <div>
                <Button className="button--big" type="submit" text="Login" />
              </div>
              <div className="auth__extra">
                <div className="auth__extra-item">
                  Yet to create an account?
                </div>
                <Link className="auth__extra-item" to={frontendRoutes.SIGN_UP}>
                  Sign up
                </Link>
              </div>
            </div>
          </form>
        </div>
      </div>
    );
  }
}

LoginPage.defaultProps = {
  startUserLogin: () => {},
  auth: false,
  userId: undefined,
};

LoginPage.propTypes = {
  startUserLogin: PropTypes.func,
  auth: PropTypes.bool,
  userId: PropTypes.number,
};

const mapStateToProps = state => ({
  auth: state.auth.isAuthenticated,
  userId: state.auth.user.id,
  error: state.auth.error,
});

const mapDispatchToProps = dispatch => ({
  startUserLogin: bindActionCreators(startUserSignIn, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(LoginPage);
