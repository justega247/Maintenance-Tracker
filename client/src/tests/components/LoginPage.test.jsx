import React from 'react';
import { shallow } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import LoginPage from '../../components/auth/LoginPage';

let startUserLogin;
let wrapper;

const createMockStore = configureMockStore([thunk]);
const store = createMockStore({});

describe('Login Page', () => {
  beforeEach(() => {
    startUserLogin = jest.fn();
    wrapper = shallow(<LoginPage.WrappedComponent
      startUserLogin={startUserLogin}
      store={store}
    />);
  });

  test('should correctly render LoginPage', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should set username on input change', () => {
    const value = 'tommy';
    const name = 'username';
    wrapper.find('input').at(0).simulate('change', {
      target: { name, value },
    });
    expect(wrapper.state(name)).toBe(value);
  });

  test('should set password on input change', () => {
    const value = 'passing';
    const name = 'password';
    wrapper.find('input').at(1).simulate('change', {
      target: { name, value },
    });
    expect(wrapper.state(name)).toBe(value);
  });

  test('should return error for invalid form submission', () => {
    wrapper.find('form').simulate('submit', {
      preventDefault: () => { },
    });
    expect(wrapper.state('errors')).not.toBe({});
    expect(wrapper).toMatchSnapshot();
  });

  test('should call startUserLogin props for valid form submission', () => {
    const validDetails = {
      username: 'allison',
      password: 'passing',
    };

    wrapper.setState(validDetails);
    wrapper.find('form').simulate('submit', {
      preventDefault: () => { },
    });
    expect(startUserLogin).toHaveBeenCalledWith(validDetails);
    expect(wrapper).toMatchSnapshot();
  });

  test('should redirect user', () => {
    const wrapper1 = shallow(<LoginPage.WrappedComponent
      startUserLogin={startUserLogin}
      store={store}
      auth
      userId={1}
    />);
    expect(wrapper1).toMatchSnapshot();
  });
});
