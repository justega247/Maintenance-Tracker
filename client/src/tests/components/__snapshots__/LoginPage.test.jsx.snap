// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page should call startUserLogin props for valid form submission 1`] = `
<div
  className="create__request"
>
  <div
    className="create__request"
  >
    <form
      className="auth__form"
      onSubmit={[Function]}
    >
      <div
        className="auth__body"
      >
        <div>
          <label
            className="label request__label"
            htmlFor="username"
          >
            Username:
            <div>
              <input
                className="form__input"
                name="username"
                onChange={[Function]}
                placeholder="enter your username"
                type="text"
                value="allison"
              />
            </div>
          </label>
        </div>
        <div>
          <label
            className="label request__label"
            htmlFor="password"
          >
            Password:
            <div>
              <input
                className="form__input"
                name="password"
                onChange={[Function]}
                placeholder="enter your password"
                type="password"
                value="passing"
              />
            </div>
          </label>
        </div>
        <div>
          <Button
            className="button--big"
            text="Login"
            type="submit"
          />
        </div>
        <div
          className="auth__extra"
        >
          <div
            className="auth__extra-item"
          >
            Yet to create an account?
          </div>
          <Link
            className="auth__extra-item"
            replace={false}
            to="/signup"
          >
            Sign up
          </Link>
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`Login Page should correctly render LoginPage 1`] = `
<div
  className="create__request"
>
  <div
    className="create__request"
  >
    <form
      className="auth__form"
      onSubmit={[Function]}
    >
      <div
        className="auth__body"
      >
        <div>
          <label
            className="label request__label"
            htmlFor="username"
          >
            Username:
            <div>
              <input
                className="form__input"
                name="username"
                onChange={[Function]}
                placeholder="enter your username"
                type="text"
                value=""
              />
            </div>
          </label>
        </div>
        <div>
          <label
            className="label request__label"
            htmlFor="password"
          >
            Password:
            <div>
              <input
                className="form__input"
                name="password"
                onChange={[Function]}
                placeholder="enter your password"
                type="password"
                value=""
              />
            </div>
          </label>
        </div>
        <div>
          <Button
            className="button--big"
            text="Login"
            type="submit"
          />
        </div>
        <div
          className="auth__extra"
        >
          <div
            className="auth__extra-item"
          >
            Yet to create an account?
          </div>
          <Link
            className="auth__extra-item"
            replace={false}
            to="/signup"
          >
            Sign up
          </Link>
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`Login Page should redirect user 1`] = `
<Redirect
  push={false}
  to="/admin-dashboard"
/>
`;

exports[`Login Page should return error for invalid form submission 1`] = `
<div
  className="create__request"
>
  <div
    className="create__request"
  >
    <form
      className="auth__form"
      onSubmit={[Function]}
    >
      <div
        className="auth__body"
      >
        <div>
          <label
            className="label request__label"
            htmlFor="username"
          >
            Username:
            <div>
              <input
                className="form__input"
                name="username"
                onChange={[Function]}
                placeholder="enter your username"
                type="text"
                value=""
              />
            </div>
          </label>
          <div
            className="form__error"
            key="username field is required."
          >
            username field is required.
          </div>
        </div>
        <div>
          <label
            className="label request__label"
            htmlFor="password"
          >
            Password:
            <div>
              <input
                className="form__input"
                name="password"
                onChange={[Function]}
                placeholder="enter your password"
                type="password"
                value=""
              />
            </div>
          </label>
          <div
            className="form__error"
            key="password field is required."
          >
            password field is required.
          </div>
        </div>
        <div>
          <Button
            className="button--big"
            text="Login"
            type="submit"
          />
        </div>
        <div
          className="auth__extra"
        >
          <div
            className="auth__extra-item"
          >
            Yet to create an account?
          </div>
          <Link
            className="auth__extra-item"
            replace={false}
            to="/signup"
          >
            Sign up
          </Link>
        </div>
      </div>
    </form>
  </div>
</div>
`;
